funsor/domains.py

```diff
@@ -1,89 +1,161 @@
 # Copyright Contributors to the Pyro project.
 # SPDX-License-Identifier: Apache-2.0
 
+import copyreg
+import functools
 import operator
-from collections import namedtuple
+import warnings
 from functools import reduce
+from weakref import WeakValueDictionary
 
 import funsor
 import funsor.ops as ops
-from funsor.util import broadcast_shape, get_tracing_state, lazy_property, quote
+from funsor.util import broadcast_shape, get_tracing_state, quote
 
 
-class Domain(namedtuple('Domain', ['shape', 'dtype'])):
-    """
-    An object representing the type and shape of a :class:`Funsor` input or
-    output.
-    """
-    def __new__(cls, shape, dtype):
-        assert isinstance(shape, tuple)
+class Domain(type):
+    def __repr__(cls):
+        return cls.__name__
+
+    def __str__(cls):
+        return cls.__name__
+
+
+class RealType(Domain):
+    _type_cache = WeakValueDictionary()
+
+    def __getitem__(cls, shape):
+        if not isinstance(shape, tuple):
+            shape = (shape,)
         # in some JAX versions, shape can be np.int64 type
         if get_tracing_state() or funsor.get_backend() == "jax":
             shape = tuple(map(int, shape))
```

Comment:
note that if we switch to `typing.Generic` we'll need to override `__getitem__` or `__class_getitem__` to do some casting.